### ISSUE TITLE
Integer arrays as buffer: OK in vanilla, problems in NG

### ISSUE THREAD

**Issue by GWRon** [bmx-ng/bmx-ng#75]
For my streamed audio I use an integer array to store stereo audio data.

It works without crackled sound in Linux but on my Windows boxes I have some cracks here and there in the audio.

Compiling the same stuff with vanilla leads to "perfect" sound output.

I think there was some adjustments in the way NG handles "arrays" (structure wise). So when passing the integer array via "byte ptr" to eg. ogg_dec (to store decoded ogg data in there) this might create some incompatibilities.

Is there a way to make the same code work on both BlitzMax incarnations or do I need to play with a `TBank` or such a thing?

How would that translate then?

Instead of defining the buffer with `new Int[ bufferlength ]` I tried a `bank.Resize( bufferlength * 2)` (EAV short after start) (and also a `bank.Resize( bufferlength * sizeof(0) )` to fetch the actual size of the integer there - it started with a random noise and then played the music but still crackling). All resulted in a hanged application and a random noise of sound.

Instead of passing `byte ptr(integerarray)` I of course used `bank.Lock()`.

Do I have to pay attention to something else?

**Comment by GWRon** [comment]
Assume the "crackling sound" hasn't to do with the buffer itself but another component in the freeaudio-ogg-decoding-chain. Something which somehow differs between vanilla and NG.
Sound playback itself with NG works (tried same soundmanager with rtAudio).

**Comment by GWRon** [comment]
Hmm, vanilla build crackles too on that notebook but works flawless on the desktop. Might that be a freeaudio-driver issue?

**Comment by GWRon** [comment]
So this is a bit odd now:
- booted linux on that "crackling windows" notebook and linux binary played sound as it should
- executed the windows build on my normal desktop linux (via wine) and no crackling sound
- executed the windows build in a (XP) VM and got no crackling sound

Means only the notebook running Windows 10 had these artifacts during playing back the audio. The audio was even borked up with a 1MB-memory-cache (which should be enough for some seconds of audio without threaded refills). The audio playback was fine when using rtAudio/maxmod2.mod for playing back the whole thing instead of utilizing freeaudio + buffer-refill code.

So ... was it my refill code? Seems not. Replaced the whole soundmanager, custom Audiostream type etc with a simple:
```BlitzMax
local s:TSound = LoadSound("sound.ogg")
PlaySound(s)
```
And the played music was "noisy/crackling" too. The testfile was a different file than in my real application (just in case the .ogg file was somehow creating the trouble).

Reading the pub.freeaudio.mod code it seems as if DirectSound was deactivated for NG but as I had similar issues with a vanilla build I am now a bit clueless (especially as it works with maxmod2.mod/rtaudio.mod).

Any ideas?

**Comment by GWRon** [comment]
Installed old v1.50 of BlitzMax and played back the same ogg file ... no cracked sounds. Renamed issue accordingly and will try to narrow it down furthermore.

**Comment by woollybah** [comment]
So what version of "vanilla" doesn't work?

**Comment by GWRon** [comment]
Ahhh found out what makes it crackling.

On windows "audio" defaults to `SetAudioDriver("DirectSound")`. So if you manually set it to `SetAudioDriver("FreeAudio")` you get this cracks in the sound playback.

**Comment by GWRon** [comment]
```BlitzMax
Framework Brl.StandardIO
Import Brl.OggLoader
Import Brl.DirectSoundAudio
Import Brl.FreeAudioAudio
Import Brl.GLMax2D

SetAudioDriver("FreeAudio") 'comment out to use DS on Win32

Graphics 500, 200
Repeat
  If KeyHit(KEY_SPACE)
    Local sound:TSound = LoadSound("your.ogg")
    PlaySound(sound)
  EndIf
  Cls
  Flip 0
Until KeyHit(KEY_ESCAPE) Or AppTerminate()
```

**Comment by woollybah** [comment]
Crackling is usually related to gaps in buffers or some such.

**Comment by GWRon** [comment]
Yes I thought the same and eg in the dsounddevice.cpp there is already an "4 x buffer spacing" comment.

But maxmods disabled the DirectSound-Driver for FreeAudio - so only "Multimedia" was left.
I then manually set `SetAudioDriver("FreeAudio Multimedia")` in a BlitzMax 1.50 (no updated modules) and it crackles there.
So it's not the DSound but the Multimedia interface creating the trouble. Will have a further look now.

**Comment by GWRon** [comment]
Ok ... think I found it:

Open up "pub.mod/freeaudio.mod/mmdevice.cpp:
```CPP
struct winmmdevice:audiodevice {
[...]
	int reset()
	{
		int sz=(iswin98())?2048:1024;	//10ms fragment size 20ms for windows98				
```
Manually setting that to 2048 removed the cracks for me - just do not know why and if a default of 20ms would work on other OS revisions too.

**Comment by GWRon** [comment]
This is the buffer size for the mixer - maybe on this notebook here some scheduler (eg Windows 10? - if so, it should crack on other machines with Windows 10 too I just do not have others here) makes it not "mix/refill" fast enough leading to this cracks here and there.

As it is a buffer size it should be safe to use `2048` - while it might occupy something a bit longer than needed (some milliseconds)

**Comment by GWRon** [comment]
Sorry for spamming now: "dsounddevice.cpp" initializes the mixer with `new mixer(DSOUNDFRAG*6)`. This is defined as  `#define DSOUNDFRAG 2048`. So with DirectSound we initialize the mixer with a size of `2048*6`. The Multimedia mixer is initialized with only `1024`. This might mean that even `2048` is not enough.
